experimental/FTheoryTools/WeierstrassModels/constructors.jl

```diff
@@ -0,0 +1,175 @@
+
+
+@doc Markdown.doc"""
+    global_weierstrass_model(f::MPolyElem{fmpq}, g::MPolyElem{fmpq}, base::AbstractNormalToricVariety)
+
+This method operates analogously to `global_weierstrass_model(base::AbstractNormalToricVariety)`.
+The only difference is that the Weierstrass sections ``f`` and ``g`` can be specified with non-generic values.
+
+# Examples
+```jldoctest
+julia> base = test_base()
+Normal toric variety
+
+julia> f = sum([rand(Int) * b for b in basis_of_global_sections(anticanonical_bundle(base)^4)]);
+
+julia> g = sum([rand(Int) * b for b in basis_of_global_sections(anticanonical_bundle(base)^6)]);
+
+julia> w = global_weierstrass_model(f, g, base)
+Global Weierstrass model over a concrete base
+
+julia> is_smooth(toric_ambient_space(w))
+false
+```
+"""
+function global_weierstrass_model(f::MPolyElem{fmpq}, g::MPolyElem{fmpq}, base::AbstractNormalToricVariety)
+    if (parent(f) != cox_ring(base)) || (parent(g) != cox_ring(base))
+        throw(ArgumentError("All Weierstrass sections must reside in the Cox ring of the base toric variety"))
+    end
+    toric_ambient_space = _ambient_space_from_base(base)
+    pw = _weierstrass_polynomial(f, g, cox_ring(toric_ambient_space))
+    Y4 = closed_subvariety_of_toric_variety(toric_ambient_space, [pw])
+    model = GlobalWeierstrassModel(f, g, pw, base, toric_ambient_space, Y4)
+    set_attribute!(model, :base_fully_specified, true)
+    return model
+end
+
+
+################################################
+# 3: Constructors over not fully specified bases
+################################################
+
+@doc Markdown.doc"""
+    global_weierstrass_model(poly_f::MPolyElem{fmpq}, poly_g::MPolyElem{fmpq}, auxiliary_base_ring::MPolyRing, d::Int)
+
+This method constructs a global Weierstrass model over a base space that is not
+fully specified. The following example illustrates this approach.
+
+# Examples
+```jldoctest
+julia> auxiliary_base_ring, (f, g, x) = QQ["f", "g", "x"];
+
+julia> w = global_weierstrass_model(f, g, auxiliary_base_ring, 3)
+Global Weierstrass model over a not fully specified base
+
+julia> weierstrass_polynomial(w)
+f*x*z^4 + g*z^6 + x^3 - y^2
+
+julia> toric_base_space(w)
+[ Info: Base space was not fully specified. Returning AUXILIARY base space.
+Normal toric variety
+
+julia> toric_ambient_space(w)
+[ Info: Base space was not fully specified. Returning AUXILIARY ambient space.
+Normal, simplicial toric variety
+
+julia> dim(toric_ambient_space(w))
+5
+```
+"""
+function global_weierstrass_model(poly_f::MPolyElem{fmpq}, poly_g::MPolyElem{fmpq}, auxiliary_base_ring::MPolyRing, d::Int)
+    if (parent(poly_f) != auxiliary_base_ring) || (parent(poly_g) != auxiliary_base_ring)
+        throw(ArgumentError("All Weierstrass sections must reside in the provided auxiliary base ring"))
+    end
+    if d <= 0
+        throw(ArgumentError("The dimension of the base space must be positive"))
+    end
+    if length([string(k) for k in gens(auxiliary_base_ring)]) < d
```

Comment:
Huh... isn't this the same as
```suggestion
    if ngens(auxiliary_base_ring) < d
```